extensions/amp-form/0.1/amp-form.js

```diff
@@ -908,32 +908,40 @@ export class AmpForm {
    * @private
    */
   handleXhrSubmitSuccess_(response) {
-    const json = /** @type {!Promise<!JsonObject>} */ (response.json());
-    return this.handleSubmitSuccess_(json).then(() => {
-      this.triggerFormSubmitInAnalytics_('amp-form-submit-success');
-      this.maybeHandleRedirect_(response);
-    });
+    return response
+      .json()
+      .then(
+        /** @type {!JsonObject} */ json => {
+          return this.handleSubmitSuccess_(json);
+        },
+        error => {
+          user().error(TAG, 'Failed to parse response JSON: %s', error);
+        }
+      )
+      .then(() => {
+        this.triggerFormSubmitInAnalytics_('amp-form-submit-success');
+        this.maybeHandleRedirect_(response);
+      });
   }
 
   /**
    * Transition the form to the submit success state.
-   * @param {!Promise<!JsonObject>} jsonPromise
+   * @param {!JsonObject} result
+   * @param {?JsonObject=} opt_eventData
    * @return {!Promise}
    * @private visible for testing
    */
-  handleSubmitSuccess_(jsonPromise) {
-    return jsonPromise.then(
-      json => {
-        this.setState_(FormState.SUBMIT_SUCCESS);
-        this.renderTemplate_(json || {}).then(() => {
-          this.triggerAction_(FormEvents.SUBMIT_SUCCESS, json);
-          this.dirtinessHandler_.onSubmitSuccess();
-        });
-      },
-      error => {
-        user().error(TAG, 'Failed to parse response JSON: %s', error);
-      }
-    );
+  handleSubmitSuccess_(result, opt_eventData) {
+    this.setState_(FormState.SUBMIT_SUCCESS);
+    return tryResolve(() => {
```

Comment:
`tryResolve` is necessary to return a rejected Promise if `renderTemplate_()` throws a synchronous error. I don't think this is possible.